src/buffer_manager.ts

```diff
@@ -425,6 +428,104 @@ export class BufferManager implements Disposable {
 
     private handleWindowChangedDebounced = debounce(this.handleWindowChanged, 100, { leading: false, trailing: true });
 
+    private async syncDocumentDirtyState(): Promise<void> {
+        const states = Array.from(this.textDocumentToBufferId.entries()).map(([doc, bufId]) => ({
+            buf: bufId,
+            modified: doc.isDirty,
+        }));
+        await this.client.lua(
+            `
+            local states = ...
+            for _, state in ipairs(states) do
+                vim.bo[state.buf].modified = state.modified
+            end
+            `,
+            [states],
+        );
+    }
+
+    private async handleSaveBuf({
+        buf,
+        bang,
+        current_name,
+        target_name,
+    }: {
+        buf: number;
+        bang: boolean;
+        current_name: string;
+        target_name: string;
+    }) {
+        // Note: workspace.save and workspace.saveAs are smart enough to handle
+        // the documents that are not a real file (e.g. untitled, output, etc.)
+        // so we can just call them directly
+
+        const document = this.getTextDocumentForBufferId(buf);
+        if (document == null) {
+            throw new Error("Received an invalid buffer id from Neovim, cannot save");
```

Comment:
generally good to mention actual values in messages, where possible.
is the id coming "from Nvim" in this case? doesn't look like it.

```suggestion
            throw new Error(`cannot save invalid Nvim buffer id: ${buf}`);
```